### PR TITLE
Reduces allocations for Multigrid solver on CPU: Multilevel structure only created when required

### PR DESCRIPTION
**Pull request by elise-palethorpe**
The multilevel structure should only be created when the matrix is initialized or altered.

cc @navidcy

### CODE REVIEW

**Comment by navidcy** [comment]
apparently autocorrect changes multigrid to multigrain... :)

File: src/Solvers/multigrid_solver.jl [hunk 1]
```diff
@@ -86,18 +87,23 @@ function MultigridSolver(linear_operation!::Function,
...
                            maxiter,
                            amg_algorithm,
+                           ml,
                            x_array,
                            b_array
                            )
 end
 
+@inline create_multilevel(::RugeStubenAMG, A) = ruge_stuben(A)
+@inline create_multilevel(::SmoothedAggregationAMG, A) = smoothed_aggregation(A; kwargs...)
```
Comment (navidcy) [inline src/Solvers/multigrid_solver.jl:106]: what's `kwargs` here?

**Review by navidcy — APPROVED** [review]
Looks good to me! If tests pass then let's merge.

File: src/Solvers/multigrid_solver.jl (hunk 1 above)
Comment (elise-palethorpe) [reply src/Solvers/multigrid_solver.jl:106]: Oops sorry, copy paste - I'll delete it

**Pull request merged**